MoreLinq/ZipImpl.cs

```diff
@@ -0,0 +1,100 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2018 Leandro F. Vieira (leandromoh). All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections;
+
+    static partial class MoreEnumerable
+    {
+        static IEnumerable<TResult> ZipImpl<T1, T2, T3, T4, TResult>(
+            IEnumerable<T1> s1,
+            IEnumerable<T2> s2,
+            IEnumerable<T3> s3,
+            IEnumerable<T4> s4,
+            Func<T1, T2, T3, T4, TResult> resultSelector,
+            int limit,
+            Func<IEnumerator[], Exception> errorSelector = null
+            )
+        {
+            IEnumerator<T1> e1 = null;
+            IEnumerator<T2> e2 = null;
+            IEnumerator<T3> e3 = null;
+            IEnumerator<T4> e4 = null;
+            var disposals = 0;
+            int calls;
```

Comment:
This should be scoped to _inside_ the while-loop, which means the same for the `GetValue` local function.

Reply:
This looks odd to me. I expect to see local functions always in the end of the outer function. 
Inside the while-loop the local function difficults the reading of the flow at all, because dont let me see if there is something beetween its end and the end of the loop.

Reply:
I understand but that's a side-effect of your design. If you don't like local functions in the middle of your code then you need to pass more of the context as arguments, like `calls`. It's much harder to reason about the data flow when the scope of `calls` needs to be moved out of the loop for the artificial purpose of having the local function at the bottom of the enclosing method's body. For example, send `calls` by reference to `GetValue` and keep the local function at the bottom as you have now:

```c#
T GetValue<T>(ref IEnumerator<T> e, ref int calls)
{
    // implementation omitted.
}
```

Now you can scope `calls` to the loop:

```c#
while (true)
{
    var calls = 0;
    var v1 = GetValue(ref e1, ref calls);
    var v2 = GetValue(ref e2, ref calls);
    var v3 = GetValue(ref e3, ref calls);
    var v4 = GetValue(ref e4, ref calls);

    if (disposals <= limit)
        yield return resultSelector(v1, v2, v3, v4);
    else
        yield break;
}
```

I still find it strange though to see a variable in a loop where _it's never used_ in the body of the loop. You have to look _inside_ `GetValue` to understand what's really going on. As I said, before, the final design might become clearer with time. For the moment, I am looking for _mechanical_ clarity. In that light, I feel it's even better to just do this:

```c#
using (e1 = s1 .GetEnumerator())
using (e2 = s2 .GetEnumerator())
using (e3 = s3?.GetEnumerator())
using (e4 = s4?.GetEnumerator())
{
    while (true)
    {
        var calls = 0;
        var v1 = GetValue(ref e1, ++calls);
        var v2 = GetValue(ref e2, ++calls);
        var v3 = GetValue(ref e3, ++calls);
        var v4 = GetValue(ref e4, ++calls);

        if (disposals <= limit)
            yield return resultSelector(v1, v2, v3, v4);
        else
            yield break;
    }
}

T GetValue<T>(ref IEnumerator<T> e, /* ref */ int calls)
{
    // call++;
    // rest omitted
}
```

So `calls` doesn't need to be passed by-reference, it can remain in the scope of the loop and it's _partially_ clear how it's being used (I say partially because we see it's being incremented unconditionally but we don't see how `GetValue` uses it).

Reply:
I see and agree that passing `calls `as argument to `GetValue` is the better we can do. you are right, it is better avoid side-effects for clarity.